src/plugins/query_enhancements/common/types.ts

```diff
@@ -13,3 +14,45 @@ export interface FetchDataFrameContext {
 }
 
 export type FetchFunction<T, P = void> = (params?: P) => Observable<T>;
+
+/**
+ * Job states are defined to mostly be compatible with EMR job run states, but aren't strictly
+ * limited to that use. "PENDING" has been replaced with "WAITING", since the underlying async API
+ * seems to use this state instead. See also: {@link parseJobState}.
+ */
+export enum JobState {
+  SUBMITTED = 'SUBMITTED',
+  WAITING = 'WAITING',
+  SCHEDULED = 'SCHEDULED',
+  RUNNING = 'RUNNING',
+  FAILED = 'FAILED',
+  SUCCESS = 'SUCCESS',
+  CANCELLING = 'CANCELLING',
+  CANCELLED = 'CANCELLED',
+}
+
+/**
+ * Convert a string to a {@link JobState} if possible. Case-insensitive. Returns `null` for
+ * invalid strings or non-strings.
+ *
+ * @param maybeState An optional string.
+ * @returns The corresponding {@link JobState} if one exists, otherwise null.
+ */
+export const parseJobState = (maybeState: unknown): JobState | null => {
+  if (!maybeState || typeof maybeState !== 'string') {
+    return null;
+  }
+  maybeState = maybeState.toUpperCase();
+  const result = Object.values(JobState).find((state) => state === maybeState);
+  return result ?? null;
+};
```

Comment:
Let's move `Object.values(JobState)` outside so we don't produce it for every call.
```
const jobStatesMap = Object.values(x).reduce(
    (acc, value) => {
      acc.set(value.toUpperCase(), value); 
      return acc;
    }, 
    new Map()
  ); 
```

```suggestion
export const parseJobState = (maybeState: unknown): JobState | undefined =>
  return jobStatesMap.get(maybeState?.toUpperCase?.());
```

Reply:
Implemented, and grabbed some low-hanging-fruit test coverage